### PR TITLE
Speed up restart writes

### PR DESCRIPTION
**Pull request by slevisconsulting**
### Description of changes
[E3SM#3163](https://github.com/E3SM-Project/E3SM/pull/3163) makes some changes to speed up restart writes.

As a first step I use E3SM#3163 function GetGlobalIndexArray in place of GetGlobalIndex so as to get global indices by array rather than looping over every element.

### Specific notes

Contributors other than yourself, if any:
@billsacks 

CTSM Issues Fixed (include github issue #):
#801 

Are answers expected to change (and if so in what way)?
No.

Any User Interface Changes (namelist or namelist defaults changes)?
No.

Testing performed, if any:
./create_test ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc -c /glade/p/cgd/tss/ctsm_baselines/ctsm1.0.dev080 -p P93300041 &	**PASS**

### CODE REVIEW

**Comment by slevisconsulting** [comment]
As far as I can tell, E3SM#3163 routine subgridRest_define_only and other similar attempts were tested and ultimately removed, so the changes that I have committed match E3SM#3163.

I have started the standard test suite hoping to see bit-for-bit same answers as baseline.

**Comment by slevisconsulting** [comment]
No failures on izumi.
One failure on cheyenne that is not among the expected:
`FAIL SMS_D_Vnuopc.f10_f10_musgs.I2000Clm50SpRsGs.cheyenne_intel.clm-default RUN time=48`
I will investigate...

**Comment by billsacks** [comment]
@slevisconsulting that is a very new test, and possibly the only one that invokes an esmf library. I wouldn't be surprised if there was some system change that is causing the failure. I would check to see if it passes from master before spending time digging into it.

**Comment by slevisconsulting** [comment]
The test that failed the first time around, passed this time:
`./create_test SMS_D_Vnuopc.f10_f10_musgs.I2000Clm50SpRsGs.cheyenne_intel.clm-default -c /glade/p/cgd/tss/ctsm_baselines/ctsm1.0.dev080 -g ctsm1.0.dev081`

**Comment by slevisconsulting** [comment]
Ready for review :-)

**Comment by slevisconsulting** [comment]
For a timing comparison, I ran this test once in master and once in the branch:
`./create_test ERS_Ld3.f09_g17.I1850Clm50BgcCrop.cheyenne_intel.clm-rad_hrly_light_res_half`
and compared walltotal stats without adding new `tstart_f` & `tstop_f` calls.

These seemed the most relevant walltotal values:
l:clm_drv_io_wrest from 14544.37 to 13304.20 => -8.53%
CPL:HISTORY from 3745.626 to 3360.475 => -10.28%
CPL:LND_RUN from 18359.34 to 17177.41 => -6.44%

@billsacks pls let me know if this makes sense, or if I have misinterpreted the stats.

**Comment by ekluzek** [comment]
I think you should run at least a few tests of the same run for the control and experiment to get some idea of the variability. Make sure that history files are being written out as well. The two relevant timers are clm_drv_io_htapes and clm_drv_io_wrest (and the timer around both is clm_drv_io.

**Comment by slevisconsulting** [comment]
@ekluzek thank you for these suggestions. I have now switched to a test that also writes history:
`./create_test ERS_Ld3.f09_g17.I2000Clm50Fates.cheyenne_intel.clm-FatesColdDef`

I have run three times in master and three in the branch and have found wide variability in the results. Too bad that the first comparison seemed so promising...

1 clm_drv_io_htapes from 2772.498 to 2413.039 => -12.97%
1 clm_drv_io_wrest from 7117.124 to 5975.481 => -16.04%
2 htapes from 1869.066 to 2671.192 => +
2 wrest from 6025.701 to 6188.046 => +
3 htapes from 2478.247 to 2296.148 => -
3 wrest from 6411.454 to 6476.287 => +

ERS1+2+3 htapes from 7,119.811 to 7,380.379 => +3.66%
ERS1+2+3 wrest from 19,554.279 to 18,639.814 => -4.68%

I think it will take a large number of runs to get statistical significance in these numbers.

@billsacks pointed out to me that E3SM#3163 states:
"The time for high-res (r0125) restart file output dropped from ~2000s to ~55s using these changes."
Is r0125 the same as our 0.125x0.125 resolution?

**Comment by billsacks** [comment]
Thanks for trying to do the timing comparison, but I'd suggest giving up on it at this point: I think it will take too much of your time. This is a nice change to have anyway for the future.

File: src/main/GetGlobalValuesMod.F90 [hunk 1]
```diff
@@ -66,6 +67,60 @@ integer function GetGlobalIndex(decomp_index, clmlevel)
...
+    use clm_varcon , only: nameg, namel, namec, namep
+    use clm_varctl , only: iulog
+    use mct_mod
+    !
+    ! Arguments 
+    integer                 , intent(in) :: bounds1
+    integer                 , intent(in) :: bounds2
+    integer, dimension(bounds1:bounds2), intent(in) :: decomp_index
+    character(len=*)        , intent(in) :: clmlevel
+    integer, dimension(bounds1:bounds2)  :: GetGlobalIndexArray
```
[the remark this answers is not in the document]
Comment (slevisconsulting) [reply src/main/GetGlobalValuesMod.F90]: I made the change to intent(in) array `decomp_index`.

GetGlobalIndexArray is declared here as the returned array and, therefore, does not need the SHR_ASSERT check, do you agree?

**Comment by slevisconsulting** [comment]
> In addition, if you haven't already, can you please do a manual cprnc of the restart file for one global resolution test, against an equivalent restart file produced from master? I ask for this because the test suite doesn't compare restart files, and I think an error in these subgrid indices on the restart files would go undetected, because I think they're just metadata that isn't actually read upon restart. When you do this, you'll need to use the `-m` flag to cprnc (which is needed when comparing files that don't have a time dimension).

@billsacks I took a shortcut by running `ncdump` on the two restart files and taking a direct `diff` on the two text files:
`diff clm2.r.asc ../../ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_191229_vyvf11/run/clm2.r.asc | more`

and this is what I got, which tells me that the files are effectively identical:
1c1
< netcdf ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_185341_9ze3np.clm2.r.0001-01-05-00000 {
---
> netcdf ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_191229_vyvf11.clm2.r.0001-01-05-00000 {
4169c4169
<               :history = "created on 01/10/20 18:58:44" ;
---
>               :history = "created on 01/10/20 19:17:24" ;
4172c4172
<               :version = "ctsm1.0.dev080-3-gea7ed9aa" ;
---
>               :version = "ctsm1.0.dev080" ;
4176c4176
<               :case_id = "ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_185341_9ze3np" ;
---
>               :case_id = "ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_191229_vyvf11" ;
634965,634966c634965,634966
<   "./ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_185341_9ze3np.clm2.h0.0001-01-05-00000.nc                                                                      ",
<   "./ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_185341_9ze3np.clm2.h1.0001-01-05-00000.nc                                                                      " ;
---
>   "./ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_191229_vyvf11.clm2.h0.0001-01-05-00000.nc                                                                      ",
>   "./ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_191229_vyvf11.clm2.h1.0001-01-05-00000.nc                                                                      " ;
634969,634970c634969,634970
<   "./ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_185341_9ze3np.clm2.rh0.0001-01-05-00000.nc                                                                     ",
<   "./ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_185341_9ze3np.clm2.rh1.0001-01-05-00000.nc                                                                     " ;
---
>   "./ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_191229_vyvf11.clm2.rh0.0001-01-05-00000.nc                                                                     ",
>   "./ERS_D_Ld6.f10_f10_musgs.I1850Clm45BgcCrop.cheyenne_intel.clm-clm50CMIP6frc.C.20200110_191229_vyvf11.clm2.rh1.0001-01-05-00000.nc

**Comment by slevisconsulting** [comment]
> I don't feel you need to rerun the full test suite, but should rerun at least some tests. I'd suggest rerunning the izumi test suite as a compromise.

Just started the test suite on izumi.

**Comment by slevisconsulting** [comment]
> > I'd suggest rerunning the izumi test suite as a compromise.
> 
> Just started the test suite on izumi.

PASS

File: src/main/GetGlobalValuesMod.F90 (hunk 1 above)
[the remark this answers is not in the document]
Comment (billsacks) [reply src/main/GetGlobalValuesMod.F90]: Yes, agreed – thanks.

**Comment by billsacks** [comment]
@slevisconsulting - Looks good now – thanks a lot! Please let me know if this is ready to be merged to master; if so, I'll do the merge.

**Comment by slevisconsulting** [comment]
@billsacks this is now ready to be merged to master. Thank you.

**Pull request merged**